core-bundle/src/Controller/ContentElement/LoginController.php

```diff
@@ -0,0 +1,186 @@
+            $template->message = $this->translator->trans('ERR.tooManyLoginAttempts', [], 'contao_default');
+        } elseif ($exception instanceof InvalidTwoFactorCodeException) {
+            $template->message = $this->translator->trans('ERR.invalidTwoFactor', [], 'contao_default');
+        } elseif ($exception instanceof AuthenticationException) {
+            $template->message = $this->translator->trans('ERR.invalidLogin', [], 'contao_default');
+        }
+
+        $redirectBack = false;
+        $pageAdapter = $this->framework->getAdapter(PageModel::class);
+
+        // Redirect to the last page visited
+        if ($model->redirectBack && $targetPath) {
+            $redirectBack = true;
+            $redirect = $targetPath;
+        }
+
+        // Redirect to the jumpTo page
+        elseif ($targetPage = $pageAdapter->findById($model->jumpTo)) {
+            $redirect = $this->generateContentUrl($targetPage, [], UrlGeneratorInterface::ABSOLUTE_URL);
+        }
+
+        $template->formId = 'tl_login_'.$model->id;
+        $template->forceTargetPath = (int) $redirectBack;
+        $template->targetPath = base64_encode($redirect);
+
+        if ($isTwoFactorInProgress) {
+            // Dispatch 2FA form event to prepare 2FA providers
+            $token = $this->container->get('security.token_storage')->getToken();
+            $event = new TwoFactorAuthenticationEvent($request, $token);
+            $this->eventDispatcher->dispatch($event, TwoFactorAuthenticationEvents::FORM);
+
+            $template->twoFactorEnabled = true;
+            $template->slabel = $this->translator->trans('MSC.continue', [], 'contao_default');
+
+            return $template->getResponse();
+        }
+
+        if ($pwResetPage = $pageAdapter->findById($model->pwResetPage)) {
+            $template->pwResetUrl = $this->generateContentUrl($pwResetPage);
+        }
+
+        $template->slabel = $this->translator->trans('MSC.login', [], 'contao_default');
+        $template->value = $lastUsername;
+        $template->autologin = $model->autologin;
+        $template->remembered = false;
+        $template->redirect = $redirect;
+
+        if ($isRemembered) {
+            $template->slabel = $this->translator->trans('MSC.verify', [], 'contao_default');
+            $template->value = $user->getUserIdentifier();
+            $template->remembered = true;
+        }
+
+        return $template->getResponse();
+    }
+
+    private function getTargetPath(Request $request): string|null
+    {
+        // If the form was submitted and the credentials were wrong, take the target path
+        // from the submitted data as otherwise it would take the current page
+        if ($request->isMethod('POST')) {
+            return base64_decode($request->request->get('_target_path'), true);
+        }
+
+        if ($request->query->has('redirect')) {
+            if ($this->uriSigner->checkRequest($request)) {
```

Comment:
What about this?

https://github.com/contao/contao/blob/e78b1d363791bd5e100641281de585a7cfa1abe1/core-bundle/contao/modules/ModuleLogin.php#L82-L83

Reply:
This is not needed. `UriSigner::checkRequest()` already does the same.